From 9b7d8ff8aef243ca27d17125772f570f5dde4a45 Mon Sep 17 00:00:00 2001
From: Nell <xianbin.zhu@spacemit.com>
Date: Sat, 19 Aug 2023 13:41:22 +0800
Subject: [PATCH 0085/1448] k1pro: enable first can-use version of cpuidle

this version can only trap to opensbi and exit it successfully,
nothing was done on the hardware.

Change-Id: I153f747f56da2d0794234839f09b2e5813243906
---
 arch/riscv/boot/dts/spacemit/k1-pro.dtsi | 84 ++++++++++++++++++++++++
 arch/riscv/configs/k1-pro_fpga_defconfig |  2 +
 2 files changed, 86 insertions(+)

diff --git a/arch/riscv/boot/dts/spacemit/k1-pro.dtsi b/arch/riscv/boot/dts/spacemit/k1-pro.dtsi
index 4bb0b3559421..5ef550877bdb 100644
--- a/arch/riscv/boot/dts/spacemit/k1-pro.dtsi
+++ b/arch/riscv/boot/dts/spacemit/k1-pro.dtsi
@@ -37,6 +37,8 @@ cpu_0: cpu@0 {
 			cpu-tlb = "1024 4-ways";
 			cpu-cacheline = "64Bytes";
 			cpu-vector = "0.7.1";
+			power-domains = <&CPU_PD0>;
+			power-domain-names = "sbi";
 
 			cpu0_intc: interrupt-controller {
 				#interrupt-cells = <1>;
@@ -59,6 +61,8 @@ cpu_1: cpu@1 {
 			cpu-tlb = "1024 4-ways";
 			cpu-cacheline = "64Bytes";
 			cpu-vector = "0.7.1";
+			power-domains = <&CPU_PD1>;
+			power-domain-names = "sbi";
 
 			cpu1_intc: interrupt-controller {
 				#interrupt-cells = <1>;
@@ -81,6 +85,8 @@ cpu_2: cpu@2 {
 			cpu-tlb = "1024 4-ways";
 			cpu-cacheline = "64Bytes";
 			cpu-vector = "0.7.1";
+			power-domains = <&CPU_PD2>;
+			power-domain-names = "sbi";
 
 			cpu2_intc: interrupt-controller {
 				#interrupt-cells = <1>;
@@ -103,6 +109,8 @@ cpu_3: cpu@3 {
 			cpu-tlb = "1024 4-ways";
 			cpu-cacheline = "64Bytes";
 			cpu-vector = "0.7.1";
+			power-domains = <&CPU_PD3>;
+			power-domain-names = "sbi";
 
 			cpu3_intc: interrupt-controller {
 				#interrupt-cells = <1>;
@@ -110,6 +118,7 @@ cpu3_intc: interrupt-controller {
 				interrupt-controller;
 			};
 		};
+
                 cpu-map {
                         cluster0 {
                                 core0 {
@@ -128,6 +137,81 @@ core3 {
                                 };
                         };
                 };
+
+		idle-states {
+			CPU_RET: cpu-ret {
+				compatible = "riscv,idle-state", "domain-idle-state";
+				idle-state-name = "riscv,cpu-ret";
+				riscv,sbi-suspend-param = <0x10000000>;
+				entry-latency-us = <20>;
+				exit-latency-us = <40>;
+				min-residency-us = <80>;
+				local-timer-stop;
+			};
+
+			CPU_NONRET: cpu-nonret {
+				compatible = "riscv,idle-state", "domain-idle-state";
+				idle-state-name = "riscv,cpu-nonret";
+				riscv,sbi-suspend-param = <0x90000000>;
+				entry-latency-us = <250>;
+				exit-latency-us = <500>;
+				min-residency-us = <950>;
+				local-timer-stop;
+			};
+
+			CLUSTER_RET: cluster-ret {
+				compatible = "domain-idle-state";
+				idle-state-name = "riscv,cluster-ret";
+				riscv,sbi-suspend-param = <0x11000000>;
+				entry-latency-us = <50>;
+				exit-latency-us = <100>;
+				min-residency-us = <250>;
+				wakeup-latency-us = <130>;
+				local-timer-stop;
+			};
+
+			CLUSTER_NONRET: cluster-nonret {
+				compatible = "domain-idle-state";
+				idle-state-name = "riscv,cluster-nonret";
+				riscv,sbi-suspend-param = <0x91000000>;
+				entry-latency-us = <600>;
+				exit-latency-us = <1100>;
+				min-residency-us = <2700>;
+				wakeup-latency-us = <1500>;
+				local-timer-stop;
+			};
+		};
+
+		power-domains {
+			CPU_PD0: cpu0 {
+				#power-domain-cells = <0>;
+				power-domains = <&CLUSTER_PD>;
+				domain-idle-states = <&CPU_RET>, <&CPU_NONRET>;
+			};
+
+			CPU_PD1: cpu1 {
+				#power-domain-cells = <0>;
+				power-domains = <&CLUSTER_PD>;
+				domain-idle-states =  <&CPU_RET>, <&CPU_NONRET>;
+			};
+
+			CPU_PD2: cpu2 {
+				#power-domain-cells = <0>;
+				power-domains = <&CLUSTER_PD>;
+				domain-idle-states = <&CPU_RET>, <&CPU_NONRET>;
+			};
+
+			CPU_PD3: cpu3 {
+				#power-domain-cells = <0>;
+				power-domains = <&CLUSTER_PD>;
+				domain-idle-states =  <&CPU_RET>, <&CPU_NONRET>;
+			};
+
+			CLUSTER_PD: cpu-cluster0 {
+				#power-domain-cells = <0>;
+				domain-idle-states = <&CLUSTER_RET>, <&CLUSTER_NONRET>;
+			};
+		};
 	};
 
 	clocks {
diff --git a/arch/riscv/configs/k1-pro_fpga_defconfig b/arch/riscv/configs/k1-pro_fpga_defconfig
index 49eb461244ec..97900fd46445 100644
--- a/arch/riscv/configs/k1-pro_fpga_defconfig
+++ b/arch/riscv/configs/k1-pro_fpga_defconfig
@@ -32,6 +32,8 @@ CONFIG_HZ_100=y
 CONFIG_RISCV_SBI_V01=y
 # CONFIG_RISCV_BOOT_SPINWAIT is not set
 CONFIG_PM=y
+CONFIG_CPU_IDLE=y
+CONFIG_RISCV_SBI_CPUIDLE=y
 CONFIG_CPU_FREQ=y
 CONFIG_CPU_FREQ_DEFAULT_GOV_USERSPACE=y
 CONFIG_CPU_FREQ_GOV_PERFORMANCE=y
-- 
2.47.0

